broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/NewPartitionTransitionImpl.java

```diff
@@ -71,67 +73,48 @@ public void updateTransitionContext(final PartitionTransitionContext transitionC
     steps.forEach(step -> step.onNewRaftRole(context, role));
 
     final ActorFuture<Void> nextTransitionFuture = concurrencyControl.createFuture();
+    final var nextTransition =
+        new PartitionTransitionProcess(steps, concurrencyControl, context, term, role);
+    nextTransitionFuture.onComplete((v, e) -> lastTransition = nextTransition);
+
+    final var ongoingTransitionFuture =
+        currentTransitionFuture == null
+            ? concurrencyControl.createCompletedFuture()
+            : currentTransitionFuture;
+
+    // For safety reasons we have to immediately replace the current transition future with the next
+    // transition future, such that we make sure that we enqueue all transitions after another.
+    //
+    // This means we will always add the next transition to the tail, this can become a chain of
+    // futures if we have many transitions after another. Ideally we would execute after the current
+    // transition only the last enqueue one. This should be implemented as soon we have time for it.
+    currentTransitionFuture = nextTransitionFuture;
 
-    concurrencyControl.run(
-        () -> {
-          if (currentTransition != null) {
-            LOG.info(
-                format(
-                    "Transition to %s on term %d requested while another transition is still running",
-                    role, term));
-            currentTransition.cancel(); // this will drop any subsequent transition steps
-
-            // schedule new transition as soon as the current step of the current transition
-            // has completed
-            concurrencyControl.runOnCompletion(
-                currentTransitionFuture,
-                (ok, error) -> cleanupLastTransition(nextTransitionFuture, term, role));
+    ongoingTransitionFuture.onComplete(
+        (nothing, error) -> {
+          context.setCurrentTerm(term);
```

Comment:
This is set too early. Context.currenntTerm/Role is used during `lastTransition.cleanUp()` to decide what needs to be done, and must at this point reflect the values of the last successful transition.

Reply:
for what we use that?

Reply:
This is one difference from the old transition. `currentTerm` and `currentRole` is set only after the transition is successful. So during the transition context.currentTerm and context.currentRole refers the role which we are  transitioning from. So in PartitionTransiitonStep#transitionTo(context, targetTerm, targetRole), we can take decisions based on if we are transition from leader to follower or follower to leader.

Reply:
Thanks, wasn't aware. I have just checked the old impl, where it was like that.

Reply:
For example in: https://github.com/camunda-cloud/zeebe/blob/ab06e27d64fcc21d1d718cca7f1899fada5dbe87/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStep.java#L93

It is used to figure out what needs to be done during a transition. Other steps are similar